### ISSUE TITLE
Add streaming service for queries

### ISSUE THREAD

**Issue by hurryabit** [digital-asset/daml#3878]
Broadly speaking, this service should do the following: You send a request similar to what the `/contracts/search` endpoint expects. The result consists of two parts: First, you get all _currently active_ contract instances matching that query. Second, you get a stream of all changes that would affect the outcome of the query if you run it again. That is, all create events that match the query and all archive events of contracts that have matched the query. Over-approximating the archive events is acceptable as long as the template id matches.

**Comment by dtanabe** [comment]
May also be useful to allow users to specify an optional starting offset if capturing _all_ events is important (and the subscriber has the means to persist the offset).

**Comment by S11001001** [comment]
> First, you get all _currently active_ contract instances matching that query. Second, you get a stream of all changes that would affect the outcome of the query if you run it again.

@hurryabit was broadly speaking, but I want to be more precise about the semantics.

First, I don't think there's a reason to separate these two parts. Supposing the obvious mapping onto ACS, then transactions since ACS: ACS represents now or some point in the past, so there is no reason that an archive couldn't immediately arrive for that initial set of contract instances. Even if ACS was guaranteed to be "now", I don't think that's sufficiently distinct from "now + 1 transaction" to draw semantic distinctions between ACS-sourced creates and transaction-sourced ones.

The use of ACS, and the way in which we tend to aggregate transactions when backpressured, does deliver a particular useful implication, best described in the negative: **archived contracts _may or may not_ appear in the stream**. This fits

> stream of all changes that would affect the outcome of the query if you run it again

as a created-and-archived contract does not appear in that delta, and moreover it delivers consistent semantics with the ACS, for which it is _also_ true that "archived contracts _may or may not_ appear in the stream".

----

Of course this semantic rule neither meets

> all create events that match the query

nor @dtanabe 

> capturing _all_ events

but these constraints seem to describe a different feature that is not quite in the spirit of "a streaming version of the ACS query endpoint", for which

> stream of all changes that would affect the outcome of the query if you run it again

is a more nice and accurate constraint.

**Comment by S11001001** [comment]
To summarize, two points to clarify with @hurryabit :

* [ ] whether the semantics I describe above are accurate, and
* [ ] to what degree offset should be supported, if any.

**Comment by lima-da** [comment]
To clarify for the current websocket work, the initial Websocket feature I put in, does NOT produce raw transaction data, it just uses `transactionClient` under the hood. It parse the transaction into Created/Archived set like this: `Contract.fromTransaction(tx)` and returns List of `Contract[+LfV](value: ArchivedContract \/ ActiveContract[LfV])` The query is supported(via initial client message), same way POST `contract/search` does, and client will only subscribe to transaction subject to that criteria. 
My origin thoughts on adding offset was pass it as a side JsObject to the query: 
```{
  "search": {
    "%templates": [
      {
        "moduleName": "Foo",
        "entityName": "A"
      },
      {
        "moduleName": "Foo",
        "entityName": "B"
      },
      {
        "moduleName": "Foo",
        "entityName": "C"
      }
    ],
    "foo": "bar"
  },
  "offset": {
    "begin": 0
  }
}
which will be finally get passed into TransactionClient
cc. @hurryabit @leo-da @stefanobaghino-da

**Comment by lima-da** [comment]
for reference, here is the current streaming message looks like:
```
{
  "transaction": [
    {
      "archived": {
        "contractId": "#499:0",
        "templateId": {
          "packageId": "ec4ff7a9c2adb530faf269fac0678b700aa0f99faa7624f3fdb97536b544830f",
          "moduleName": "Iou",
          "entityName": "Iou"
        },
        "witnessParties": [
          "Alice"
        ]
      }
    },
    {
      "created": {
        "observers": [],
        "agreementText": "",
        "signatories": [
          "Alice"
        ],
        "contractId": "#10943:1",
        "templateId": {
          "packageId": "ec4ff7a9c2adb530faf269fac0678b700aa0f99faa7624f3fdb97536b544830f",
          "moduleName": "Iou",
          "entityName": "Iou"
        },
        "witnessParties": [
          "Alice"
        ],
        "argument": [
          "Alice",
          "Alice",
          "USD",
          "10.0",
          []
        ]
      }
    },
    {
      "created": {
        "observers": [],
        "agreementText": "",
        "signatories": [
          "Alice"
        ],
        "contractId": "#10943:2",
        "templateId": {
          "packageId": "ec4ff7a9c2adb530faf269fac0678b700aa0f99faa7624f3fdb97536b544830f",
          "moduleName": "Iou",
          "entityName": "Iou"
        },
        "witnessParties": [
          "Alice"
        ],
        "argument": [
          "Alice",
          "Alice",
          "USD",
          "190.0",
          []
        ]
      }
    }
  ]
}
```

**Comment by hurryabit** [comment]
@S11001001 _If_ I understand you correctly, the semantics you describe is what I intended. When I send a streaming query, I want a snapshot of the ACS (restricted to the query) at some point in time, preferrable very close to now. After that, I want a stream of (possibly aggregated) transactions such that whenever I apply one of these transactions to my current snapshot of the ACS, I get a newer, more current snapshot of the ACS. (I'm fine if the first transaction is a list of creates that makes up the current ACS.) My goal is to maintain a snapshot of the ACS on the client-side.

My inclination is to not support offsets at all. I consider them an expert use case, which is already covered by the gRPC API. @dtanabe can you please confirm that we agreed on this. :)

**Comment by S11001001** [comment]
@hurryabit Yes, that precisely matches up with my understanding. Thank you.

**Comment by da-tanabe** [comment]
In further discussions within the DABL team, one use case for offsets came up. When deploying this component, we are going to want to horizontally scale this component as we are already running into cases where we overwhelm the gRPC API with too many concurrent connections. Without offsets, unless we implement stickiness (and even in the case of a component going down and coming up again, or transient network failures), each disconnect of a websocket will require dumping all local state on the client and refreshing from the beginning of time. This is wasteful at best, or may force a full page reload of a front-end client at worst. If the frontend could simply remember the last received offset, it could reconnect a disconnected websocket without too much drama.

So @hurryabit , we did talk about it and I couldn't think of a "non-expert" usecase at the time, but simply supporting cheap reconnects of a websocket doesn't seem expert…unless there is alternate way of supporting this?

**Comment by hurryabit** [comment]
@dtanabe I think I don't fully understand what you mean by "refreshing from the beginning of time". What you'll get after a reconnect are all _active_ contracts rather than the whole history of how you got there. Do you think the active contract set for a single user will be too big in the foreseeable future?

**Comment by da-tanabe** [comment]
My concern is that multiple copies of the HTTP JSON API may not necessarily be in sync with each other, so a reconnect actually ends up throwing you back in time ACS-wise.

**Comment by hurryabit** [comment]
@dtanabe I don't see how you would go back in time. The first "frame" you get after reconnecting is the current state of the ACS. That won't be any older than what you had before.

**Comment by da-tanabe** [comment]
Consider two instances of HTTP JSON API running on different servers A and B over the same ledger (again, for high-availability/resiliency purposes). Server A _happens_ to be at transaction 100 at a point in time, and another is at transaction 125—hiccups happen frequently in infrastructure, and maybe Server A is temporarily unable to reach the ledger while Server B is able to.

Then suddenly Server B is killed, causing all clients to reconnect to Server A. Suddenly all clients that were formerly connected to Server B find themselves 25 transactions in the past. Now this _will_ eventually catch up to the eventual notion of HEAD ACS, but the intermediate state presented to the user could potentially be very confusing, and also lead to erroneous double spends. And the client has no tools to detect that this scenario actually happens.

**Comment by leo-da** [comment]
I tend to agree with @da-tanabe... if we are to run HTTP JSON API behind a load balancer we need to introduce a bookmark concept, ledger offset in our case. And the HTTP JSON API needs to be idempotent, which we did not design for.

**Comment by S11001001** [comment]
The behavior described by @da-tanabe is consistent with the results of continuously polling the query service and observing deltas, and therefore attempting to elaborate semantics to avoid certain undesirable situations is outside the scope of this ticket.

The implication of JSON API always only respecting "now" is that the observation of the ledger is only consistent within a single request. We traditionally declare one exception to that: the results of a successful create/exercise can be observed on a sequentially-made request. For example, if you create, wait for 200, and then query, you'll see that contract in the query results (assuming some other process hasn't archived it in the meantime).

Now, that perfectly intuitive, easy-to-satisfy promised exception can hardly be made if you postulate that your second observation is actually made at the distance of a few light-minutes from the first observation, which is exactly the scenario entailed by the HA architecture @da-tanabe describes; consider that the create is made to server B, and the query is then immediately made to server A.

If we cannot preserve _that_ simple invariant under the scenario that @da-tanabe describes, I don't see how one could reasonably expect a significantly less simple invariant to hold under the same scenario.

In other words, because JSON API is always "now", it cannot respect the decision of clients to transparently reconnect as @da-tanabe describes and simply assume that the results immediately delivered (which is not a well-defined concept in this websocket endpoint anyway!) are consistent in some additionally promised way with prior requests. If you've designed your server architecture such that "later" might actually mean "earlier", you have to respect that uncertainty on the client side, especially when going beyond the one-request consistency strictures offered by JSON API.

**Comment by meiersi-da** [comment]
> > First, you get all _currently active_ contract instances matching that query. Second, you get a stream of all changes that would affect the outcome of the query if you run it again.
> 
> @hurryabit was broadly speaking, but I want to be more precise about the semantics.
> 
> First, I don't think there's a reason to separate these two parts. Supposing the obvious mapping onto ACS, then transactions since ACS: ACS represents now or some point in the past, so there is no reason that an archive couldn't immediately arrive for that initial set of contract instances. Even if ACS was guaranteed to be "now", I don't think that's sufficiently distinct from "now + 1 transaction" to draw semantic distinctions between ACS-sourced creates and transaction-sourced ones.
> 
> The use of ACS, and the way in which we tend to aggregate transactions when backpressured, does deliver a particular useful implication, best described in the negative: **archived contracts _may or may not_ appear in the stream**. This fits
> 
> > stream of all changes that would affect the outcome of the query if you run it again
> 
> as a created-and-archived contract does not appear in that delta, and moreover it delivers consistent semantics with the ACS, for which it is _also_ true that "archived contracts _may or may not_ appear in the stream".
> 
> Of course this semantic rule neither meets
> 
> > all create events that match the query
> 
> nor @dtanabe
> 
> > capturing _all_ events
> 
> but these constraints seem to describe a different feature that is not quite in the spirit of "a streaming version of the ACS query endpoint", for which
> 
> > stream of all changes that would affect the outcome of the query if you run it again
> 
> is a more nice and accurate constraint.

@S11001001 : this semantics doesn't clarify whether a client gets the ability to observe transaction boundaries. If you think about a client consuming a stream to build a table of balances, then it is important to only display that table to a user on transaction boundaries. Otherwies, a transfer of cash would result in money vanishing from the table of balances if displayed right after the `Archive` event on the original contract that is being transferred.

(I admit I have only skimmed the above convo, but it seems to me that this is a problem that should be covered, but is not yet. Is it?)

**Comment by S11001001** [comment]
@meiersi-da The semantics discussed previously don't cover that at all. I can tell you that it is guaranteed that:

1. A response block with an offset marker (#4593) is guaranteed to consist of one or more whole transactions;
2. conversely, a response block without an offset is not defined in relation to any transaction or ordering.

In your hypothetical table of balances, the latter will only occur when the stream is returning the initial responses, i.e. the table isn't "loaded" yet. Following that, while returning live data, the latter will never occur. This fact is specified in documentation.

The relationship between offset-containing response blocks and transactions is not specified; however, I believe the unstated implication of the presence of an offset is that _the response block's end coincides with a transaction boundary_, which prevents the confusing behavior in your example.

**Comment by meiersi-da** [comment]
Thanks for the explanations @stephencompall-DA . Glad you've already thought about the issue. Is there a particular reason why the relationship is not specified and mandated?

**Comment by S11001001** [comment]
@meiersi-da There are three reasons:

1. The WebSocket documentation is quite weighty with detail already (see especially ["Some notes on behavior"](https://docs.daml.com/0.13.54/json-api/index.html#contracts-query-stream)), so anything that allows us to elide detail is welcome.
2. As a rule, I would rather not restate ledger semantics in the JSON API documentation, the relevant one here being "offsets occur at transaction boundaries".
3. We have deliberately avoided having transactions be a first-class concept in the JSON API (notwithstanding a recent reference in create-and-exercise docs that maybe we should revisit), so if it is possible to support the underlying thing you are really concerned about without mentioning transactions at all, all the better.

I have tried to state only the essentials in the documentation, e.g. the behavior of the offset field and what happens if you pass it back in as an argument, the presence of phantom archives, the synthetic "liveness" indicator. Even then, essential details like "archived contracts may or may not appear in the stream" needed multiple rounds of discussion to clarify.

I don't have a good sense of what details and implications are important enough to give the same treatment, but perhaps this is one of them, or you might have ideas of other semantic areas we should cover in the documentation.

**Comment by meiersi-da** [comment]
Thanks for the detailed response. I understand this is judgment call, and I'm happy to leave that one to you.


While reading your notes I've also stumbled upon:
> Within a given array, if an archived and created refer to contracts with the same template ID and contract key, the archived is guaranteed to occur before the created.

Wouldn't this be the reverse: a `create id1` occur before the `archive id2` ?

**Comment by hurryabit** [comment]
> Wouldn't this be the reverse: a `create id1` occur before the `archive id2` ?

@meiersi-da Whenever a block of transaction we're shipping to the client would contain both the create and the archive event for a single _contract id_ we don't ship either of them. (You can think of the archive canceling out the create.) Thus, if you see a create and an archive for one _contract key_, the archive must be the archive of the old contract the key pointed to and the create must be a subsequent creation of a new contract with the same key. Does that explain what's going on?

**Comment by meiersi-da** [comment]
OK. That makes sense, but does not seem to match what I find under "Some notes on behaviour" at the bottom of this section: https://docs.daml.com/0.13.54/json-api/index.html#contracts-query-stream

It seems we are all aligned on the semantics, and the matter is how to inform our users well to make the productive users. The above of what I've raised looks like a typo.

**Comment by hurryabit** [comment]
@meiersi-da I can't spot the inconsistency. Are you referring to
> The output stream has the same format as the output from the Contracts Query Stream. We further guarantee that for every archived event appearing on the stream there has been a matching created event earlier in the stream.

**Comment by S11001001** [comment]
> Wouldn't this be the reverse: a `create id1` occur before the `archive id2` ?

@meiersi-da No, the order in the documentation is correct. This is equivalent to the rule from #4354 :

> whenever a block of events contains a create and an archive event affecting the same contract key, they are shipped in order

Possibly we should modify (5) to say "notwithstanding (3)", but a little more nicely.

**Comment by meiersi-da** [comment]
Thanks @stephencompall-DA and @hurryabit for your patience. I've reread the docs and realized that I've shortened "same template ID and contract key" to "same contract id" in my head when reading it. The docs are correct.

**Comment by hurryabit** [comment]
You read it as "same contract ~key and template~ id" then? 😉
